GetMountRefs fixed to handle corrupted mounts by treating it like an unmounted volume

davidz627 opened the pull request:
Fixes: #73513

/sig storage
/kind bug
/assign @msau42 @saad-ali 

```release-note
NONE
```

davidz627 committed:
GetMountRefs fixed to handle corrupted mounts by treating it like an
unmounted volume

msau42 commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1003,10 +1003,11 @@ func (mounter *Mounter) SafeMakeDir(subdir string, base string, perm os.FileMode
 }
 
 func (mounter *Mounter) GetMountRefs(pathname string) ([]string, error) {
-	if _, err := os.Stat(pathname); os.IsNotExist(err) {
```
does windows also need fixing since the original change also added a stat check for mount_windows?  cc @jingxu97 @andyzhangx

jingxu97 replied:
looks like PathExists in the mount_helper does not consider windows OS. We could add a windows version of PathExists or add windows handling inside of PathExists?

davidz627 replied:
I am not sure which windows errors are because of corrupted mount that correspond with unix errors `ENOTCONN`, `ESTALE`, `EIO`.

Windows error list here: https://github.com/golang/sys/blob/master/windows/types_windows.go#L9

msau42 replied:
Note that pkg syscall and sys are different

davidz627 replied:
right, pkg `sys` seems to imply that the same `syscall.Errno` codes mean different things in windows vs. unix.
And `IsCorruptedMnt` seems to be using the unix interpretation of the error codes so I would think we need a Windows specific version of `PathExists` and `IsCorruptedMnt`.

andyzhangx replied:
> looks like PathExists in the mount_helper does not consider windows OS. We could add a windows version of PathExists or add windows handling inside of PathExists?

whether `PathExists` works on windows os depends on whether `IsCorruptedMnt` works on windows, it at least could compile on windows. Since the original issue is related to XFS mount issue, and windows does not support that, I think use same fix on windows would be ok.
@davidz627 could you do same fix on `windows_mount.go`? 
BTW, as I implemented `GetMountRefs` func in the beginning, you could see it's not reliable since there is no place to query all mount points on Windows, different from Linux.
https://github.com/kubernetes/kubernetes/blob/4fa5ece62ce3475d851c184bb623e6220fbc4950/pkg/util/mount/mount_windows.go#L497-L505

davidz627 replied:
i've updated the windows `GetMountRefs` as well. I believe the errcodes are different for windows though. Therefore I have removed the `IsCorruptedMnt` check as I believe it would be just checking "random" error codes and may mask real issues.

Since XFS mount issue doesn't exist on windows anyway this is not a necessary addition.

davidz627 commented:
/assign @jingxu97 
@msau42 
I've updated this PR with windows changes. PTAL

davidz627 commented:
/retest

jingxu97 commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1003,10 +1003,11 @@ func (mounter *Mounter) SafeMakeDir(subdir string, base string, perm os.FileMode
-	} else if err != nil {
-		return nil, err
+	} else if pathErr != nil {
+		return nil, fmt.Errorf("Error checking path: %v", pathErr)
```
do we need to print out the path or the pathErr should already include it?

davidz627 replied:
resolved

jingxu97 commented on pkg/util/mount/mount_windows.go:
```diff
@@ -496,10 +496,11 @@ func getAllParentLinks(path string) ([]string, error) {
 func (mounter *Mounter) GetMountRefs(pathname string) ([]string, error) {
-	if _, err := os.Stat(normalizeWindowsPath(pathname)); os.IsNotExist(err) {
+	pathExists, pathErr := PathExists(normalizeWindowsPath(pathname))
+	if !pathExists {
```
how about checking IsCorruptedMnt? This function should work for both windows and linux?

davidz627 replied:
https://github.com/kubernetes/kubernetes/pull/74625#discussion_r262214952

technically it "works" as in it compiles, but I think it is checking the incorrect error codes. Since there is no `XFS` on windows anyway we do not need to check for those types of corrupted mounts anyway

jingxu97 commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1003,10 +1003,11 @@ func (mounter *Mounter) SafeMakeDir(subdir string, base string, perm os.FileMode
 func (mounter *Mounter) GetMountRefs(pathname string) ([]string, error) {
-	if _, err := os.Stat(pathname); os.IsNotExist(err) {
+	pathExists, pathErr := PathExists(pathname)
+	if !pathExists || IsCorruptedMnt(pathErr) {
```
maybe add a log for IsCorruptedMnt for debugging?

davidz627 replied:
resolved

msau42 commented:
It's not just xfs, nfs or any other type of remote network plugin can have stale mounts

andyzhangx commented:
/test pull-kubernetes-cross

davidz627 commented:
Hi @andyzhangx,

Based on @msau42's comment it looks like we still want to check for corrupted mounts on Windows. Wondering if you have the expertise to determine which error codes are corrupted mounts for windows and if you could write a Windows specific `IsCorruptedMntWindows` function. I am fine with you just cherry-picking my commit and building a new PR on top of it - or you can just send me the code for a `IsCorruptedMntWindows`, either way would work for me.

jingxu97 commented:
maybe this link https://mariadb.com/kb/en/library/operating-system-error-codes/ can help to find to error code for windows vs. Linux?

andyzhangx commented:
hi @davidz627 I think we could use the exact same `IsCorruptedMnt` on Windows, I searched the code, same error code exists in windows golang code, you could see below 3 error code is the same on Windows:
https://github.com/kubernetes/kubernetes/blob/596a48dd64bcaa01c1d2515dc79a558a4466d463/pkg/util/mount/mount_helper.go#L123

davidz627 commented:
@andyzhangx I'm not sure that's true.

Take this for example:
`underlyingError == syscall.EIO`
`EIO             = Errno(0x5)`

On Unix when we try to `Stat` the mount and we have an I/O error -> the kernel will emit errno `0x5` which will be correctly interpreted as `EIO` in this case.
However, if this was Windows, and we try to stat the mount and the error is *Access Denied* because of some permissions -> the kernel will emit errno `0x5` which will be mistakenly interpreted as `EIO` in this case.

That is the analysis from my understanding of how errno's are emitted. However, I may have some incorrect assumptions, feel free to correct me if I am wrong.

davidz627 commented:
/test pull-kubernetes-cross

andyzhangx commented:
@davidz627 I found a same unmount issue on Windows: https://github.com/kubernetes/kubernetes/issues/75025
Let me find a way to fix this.

Maybe go ahead with this PR and I will send another PR to fix that issue on Windows since it's close to code freeze already.

andyzhangx commented:
Update: as I have checked the azure file(SMB protocol) [invalid issue](https://github.com/kubernetes/kubernetes/issues/75025) on Windows, the errror is like following:
```
underlyingError = "The user name or password is incorrect"
```
It does not belong to any one of errors listed in https://golang.org/pkg/syscall/?GOOS=windows&GOARCH=amd64#Errno
Also, even if it's on Linux, how do we know only these 3 errors are regarded as orrupted mount point:
```
underlyingError == syscall.ENOTCONN || underlyingError == syscall.ESTALE || underlyingError == syscall.EIO
```

If you take a look at the above error checking, finally it will only do error string comparison, so when error message changes, it won't work:
```
syscall.ENOTCONN = transport endpoint is not connected
syscall.ESTALE = stale NFS file handle
syscall.EIO = input/output error
```

There are other file errors could still lead to this unmount issue. Thus, can we just write `IsCorruptedMnt` on Windows like this @davidz627 , I have tested, issue#75025 could be fixed by this implementation:
```
func IsCorruptedMnt(err error) bool {
	return err != nil
}
```

andyzhangx commented:
@davidz627 could you also add this patch for smb mount error(errorunderlyingError = `permission denied`), I guess there are still a lot of errors missing in this `IsCorruptedMnt` on Linux:
```
diff --git a/pkg/util/mount/mount_helper.go b/pkg/util/mount/mount_helper.go
index 696843a..9fd8ac0 100644
--- a/pkg/util/mount/mount_helper.go
+++ b/pkg/util/mount/mount_helper.go
@@ -120,5 +122,5 @@ func IsCorruptedMnt(err error) bool {
                underlyingError = pe.Err
        }

-       return underlyingError == syscall.ENOTCONN || underlyingError == syscall.ESTALE || underlyingError == syscall.EIO
+       return underlyingError == syscall.ENOTCONN || underlyingError == syscall.ESTALE || underlyingError == syscall.EIO || underlyingError == syscall.EACCES
 }
```

davidz627 commented:
I think:
> func IsCorruptedMnt(err error) bool {
> 	return err != nil
> }
is too big of a hammer. There may be other types of errors that we would actually want to surface instead of burying them all.

I have added `underlyingError == syscall.EACCES` to the list of underlying errors as per @andyzhangx's analysis.

I will let @andyzhangx follow up with a Windows specific PR. As it stands there is no `IsCorruptedMnt` checking for `mount_windows.go` but I have added a comment with a TODO to fix that.

davidz627 commented:
/priority important-soon
ping @jingxu97 @msau42 for lgtm and approval

jingxu97 commented on pkg/util/mount/nsenter_mount.go:
```diff
@@ -344,12 +344,11 @@ func (mounter *NsenterMounter) SafeMakeDir(subdir string, base string, perm os.F
+	if !pathExists || IsCorruptedMnt(pathErr) {
 		return []string{}, nil
+	} else if pathErr != nil {
+		return nil, fmt.Errorf("Error checking path: %v", pathErr)
```
Also add path into the error message like others?

jingxu97 commented:
/approve

andyzhangx commented:
@davidz627 thanks, I will work out a windows fix soon.

jingxu97 commented:
/lgtm

andyzhangx reviewed (approved):
/lgtm

andyzhangx commented:
BTW, would you also cherry-pick to older release? @davidz627
